### ISSUE TITLE
Invalid scale reference causes domain pollution (for multi-field data ref domain)

### ISSUE THREAD

**Issue by ericsoco** [vega/vega#1166]
In the following spec, the `fill` spec for the right axis causes an `undefined` value to appear in the `color` scale's domain. Verified in the REPL by viewing the rendered legend and also by logging `VEGA_DEBUG.view._runtime.scales.color.value.domain()` (returns `["A", "B", "val", undefined]`).

Removing the right axis `fill` spec fixes the problem. The `fill` spec in question is invalid, as the axis encoding must map to a single color; the same spec used e.g. within a mark encoding does not trigger this issue.

Not a problem for me now that I understand I have an incorrect usage I need to fix, but it took me a while to figure out why my domain was rendering an `undefined` value...

```
{
  "$schema": "https://vega.github.io/schema/vega/v3.0.json",
  "width": 400,
  "height": 300,
  "data": [
    {
      "name": "dataset",
      "values": [
        {"bin": "A", "val": 10},
        {"bin": "B", "val": 20},
        {"bin": "A", "val": 30},
        {"bin": "B", "val": 40}
      ]
    }
  ],
  "scales": [
    {
      "name": "y",
      "type": "linear",
      "domain": {
        "data": "dataset",
        "field": "y_out"
      },
      "zero": true,
      "range": "height"
    },
    {
      "name": "color",
      "type": "ordinal",
      "domain": {
        "fields": [
          {"data": "dataset", "field": "bin"},
          ["val"]
        ]
      },
      "range": ["#DDB27C", "#11939a", "#11939a"]
    }
  ],
  "axes": [
    {
      "orient": "left",
      "scale": "y",
      "title": "left side",
      "encode": {
        "title": {
          "update": {
            "fill": {
              "signal": "scale('color', 'val')"
            }
          }
        }
      }
    },
    {
      "orient": "right",
      "scale": "y",
      "title": "right side",
      "encode": {
        "title": {
          "update": {
            "fill": {
              "scale": "color",
              "field": "bin"
            }
          }
        }
      }
    }
  ],
  "legends": [
    {
      "fill": "color"
    }
  ]
}
```

**Comment by jheer** [comment]
Thanks @ericsoco! I believe @domoritz and I ran into this issue before. The underlying cause is that d3's ordinal scale automatically extends the domain when previously unseen values are passed through the scale function. While useful in many cases, this side-effect behavior is a problem here.

We can look into setting the ordinal scale's `unknown` value to see if we can prevent this issue in the future.

**Comment by domoritz** [comment]
If it is useful, would it be better to display a warning in this case but leave the functionality as it is?

**Comment by ericsoco** [comment]
Yikes, really? Seems like reads should be idempotent...I can understand the utility of that behavior but it's pretty unexpected IMO. I do see it [in the docs](https://github.com/d3/d3-scale#_ordinal) now, thanks for the explanation.

I think @domoritz is right -- it's possible you have consumers expecting this behavior and would be a breaking change to modify it. Probably best to display a warning and also expose a value on Vega's ordinal scale spec that allows modifying d3's ordinal.unknown, or at least turning off implicit domain construction.

**Comment by jheer** [comment]
Right now, one can define an ordinal scale with a range (e.g., category10) and no corresponding domain definition. The domain is then populated as values are passed through the scale function. As we have supported this since Vega 1, I'd be hesitant to break this feature. On the other hand, if a domain _is_ explicitly defined within the spec, it seems reasonable to me to disable automatic domain extension when new values are encountered.

**Comment by ericsoco** [comment]
Ha, I win by one second ;)

<img width="387" alt="screen shot 2018-02-21 at 10 48 36 am" src="https://user-images.githubusercontent.com/1127259/36499000-dae32162-16f4-11e8-9f4e-20688140c2dd.png">

**Comment by jheer** [comment]
Fixed in vega-encode, released with Vega 3.2!
